ethereum/core/src/main/java/tech/pegasys/teku/core/epoch/status/ValidatorStatuses.java

```diff
@@ -0,0 +1,139 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core.epoch.status;
+
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root_at_slot;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.operations.AttestationData;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.datastructures.state.Checkpoint;
+import tech.pegasys.teku.datastructures.state.Validator;
+import tech.pegasys.teku.datastructures.util.AttestationUtil;
+import tech.pegasys.teku.datastructures.util.BeaconStateUtil;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+
+public class ValidatorStatuses {
+  private final List<ValidatorStatus> statuses;
+  private final TotalBalances totalBalances;
+
+  private ValidatorStatuses(
+      final List<ValidatorStatus> statuses, final TotalBalances totalBalances) {
+    this.statuses = statuses;
+    this.totalBalances = totalBalances;
+  }
+
+  public static ValidatorStatuses create(final BeaconState state) {
+    final SSZList<Validator> validators = state.getValidators();
+
+    final UInt64 currentEpoch = BeaconStateUtil.get_current_epoch(state);
+    final UInt64 previousEpoch = BeaconStateUtil.get_previous_epoch(state);
+
+    final List<ValidatorStatus> statuses =
+        validators.stream()
+            .map(validator -> ValidatorStatus.create(validator, previousEpoch, currentEpoch))
+            .collect(Collectors.toCollection(() -> new ArrayList<>(validators.size())));
+
+    processAttestations(statuses, state, previousEpoch, currentEpoch);
+
+    final TotalBalances totalBalances = TotalBalances.create(statuses);
+
+    return new ValidatorStatuses(statuses, totalBalances);
+  }
+
+  private static void processAttestations(
+      final List<ValidatorStatus> statuses,
+      final BeaconState state,
+      final UInt64 previousEpoch,
+      final UInt64 currentEpoch) {
+    Stream.concat(
+            state.getPrevious_epoch_attestations().stream(),
+            state.getCurrent_epoch_attestations().stream())
+        .forEach(
+            attestation -> {
+              final AttestationData data = attestation.getData();
+
+              final AttestationUpdates updates = new AttestationUpdates();
+              final Checkpoint target = data.getTarget();
+              if (target.getEpoch().equals(currentEpoch)) {
+                updates.currentEpochAttester = true;
+                updates.currentEpochTargetAttester =
+                    matchesEpochStartBlock(state, currentEpoch, target.getRoot());
+              } else if (target.getEpoch().equals(previousEpoch)) {
```

Comment:
if it's not current or previous, should it be an IllegalState? theres if, else if, no else...

Reply:
I'm in two minds about this.  If it's not either, there's no rewards and we can just ignore the attestation so the current code isn't wrong.  It shouldn't ever happen because the attestations in a state only come from blocks and the block validation prevents attestations that aren't from current or previous epoch from appearing.  If that ever changed we wouldn't necessarily want this to throw an exception.

Given that we're iterating over the combination of `getPrevious_epoch_attestations` and `getCurrent_epoch_attestations` I'm going to lean towards not having the else unless there are particularly strong objections.

Reply:
I was just curious of the reasoning because its a little uncommon so I figured I'd ask the question - happy with the answer.

Reply:
It was an excellent question. :)